### ISSUE TITLE
[naming-convention] enumMembers should be default PascalCase

### ISSUE THREAD

**Issue by Mister-Hope** [typescript-eslint/typescript-eslint#7879]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Description

Code like this fails with the default option:

```ts
const enum SearchItemType {
  Page = 0,
  ID = 1,
}

const enum SearchIndexType {
  Title = 1,
  Heading = 2,
  Text = 3,
  Image = 4,
  Card = 5,
  Doc = 6,
}
```

However, I think using PascalCase for enum member should be the default option, as this is the default syntax in TypeScript official docs, blogs and examples.

E.g.: 
- https://www.typescriptlang.org/docs/handbook/enums.html
- https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#all-enums-are-union-enums

---

## Other things to report and could be a feature request

The problem is affecting me because we are not providing a keyword to enable the default naming convention and "make some tweaks over it". I.E.: In this case, I can not simply add the green line below, instead I have to redeclare the whole "option". That is not convinient for me accross projects.

### Impacted Configurations

```diff
// the default config is similar to ESLint's camelcase rule but more strict
const defaultOptions: Options = [
  {
    selector: 'default',
    format: ['camelCase'],
    leadingUnderscore: 'allow',
    trailingUnderscore: 'allow',
  },

  {
    selector: 'import',
    format: ['camelCase', 'PascalCase'],
  },

  {
    selector: 'variable',
    format: ['camelCase', 'UPPER_CASE'],
    leadingUnderscore: 'allow',
    trailingUnderscore: 'allow',
  },

  {
    selector: 'typeLike',
    format: ['PascalCase'],
  },

+  {
+    selector: 'enumMember',
+    format: ['PascalCase'],
+  },
];
```

**Comment by JoshuaKGoldberg** [comment]
Some codebases use camelCase for enums, some PascalCase. I suspect we should allow both for the default convention.

**Comment by bradzacher** [comment]
I'm not sure about this, tbh.
This has been the default since the rule was released (so it's been this way for a few years now) and this is the first time we've had this asked.

Either that means one of:
- not many people use the rule
- not many people use the rule with the defaults
- people use the rule with defaults and are happy with camelCase for enum names

Could do a sourcegraph query of public githubs to roughly determine which it is. If it's the 3rd then making this change would be bad as it would allow inconsistency in codebases - IMO you want one or the other in a codebase - not both.

**Comment by Mister-Hope** [comment]
I agree with @JoshuaKGoldberg that camelCase should be added for backward compatibility reasons.

And for the casing, I searched npm [homepage  Popular libraries](https://www.npmjs.com/) and [fronend packages sorted by popularity](https://www.npmjs.com/search?q=keywords%3Afront-end&page=0&ranking=popularity)

I omit any package which does not have built-in typescript (even it as `@types/xxx` package), also a lot of package does not use enum such as React.

1. Vue: using  PascalCase and small ammount of UPPERCASE (shims them to single char in prod) <https://github.com/search?q=repo%3Avuejs%2Fcore%20enum&type=code>
2. axios: using PascalCase <https://github.com/axios/axios/blob/f7adacdbaa569281253c8cfc623ad3f4dc909c60/index.d.cts#L138>
3. Preact: camelCase <https://github.com/preactjs/preact/blob/9a9967d4edab22581e4f9245741e62452d4b43d2/src/internal.d.ts#L4>
4. Vite: both camelCase and PascalCase <https://github.com/search?q=repo%3Avitejs%2Fvite%20enum%20language%3ATypeScript%20&type=code>
5. Webpack: camelCase <https://github.com/search?q=repo%3Awebpack%2Fwebpack%20enum%20language%3ATypeScript%20&type=code>
6. Nest.js: UPPERCASE <https://github.com/search?q=repo%3Anestjs%2Fnest%20enum%20language%3ATypeScript%20&type=code>
7. Grape.js: UPPERCASE camelCase and PascalCase <https://github.com/search?q=repo%3AGrapesJS%2Fgrapesjs%20enum%20language%3ATypeScript%20&type=code>
8. Cheerio: PascalCase <https://github.com/search?q=repo%3Acheeriojs%2Fcheerio%20enum%20language%3ATypeScript%20&type=code>
9.  Babel: UPPERCASE camelCase and PascalCase <https://github.com/search?q=repo%3Ababel%2Fbabel+enum+language%3ATypeScript+&type=code>
10. TypeScript: UPPERCASE camelCase and PascalCase, but >80% PascalCase <https://github.com/search?q=repo%3Amicrosoft%2FTypeScript+enum+language%3ATypeScript+&type=code>
11. Angular: PascalCase mostly, only a small ammount of camelCase  <https://github.com/search?q=repo%3Aangular%2Fangular%20enum%20language%3ATypeScript%20&type=code>
12. VSCode: PascalCase mostly <https://github.com/search?q=repo%3Amicrosoft%2Fvscode%20enum%20language%3ATypeScript%20&type=code>

For the above repo, only Webpack and Preact is using camelCase enum. Others are all using PascalCase mostly.

A lot of repo includes camelCase (or lowercase) to represent the same keyword strings, e.g.:

```ts
const enum Log {
  debug = 'debug',
  info = 'info',
  // ...
}
```
  
So I still inisit my point @bradzacher, at least PascalCase should be added to defaults.

**Comment by Mister-Hope** [comment]
Another reason according to my usage, using PascalCase for enum makes it easier to recognize them through codes:

```ts
// Probably a class property
Abc.def
// TypeScript enum / const enum
Abc.Def
// object property
abc.def
```

**Comment by Josh-Cena** [comment]
I also agree to add PascalCase by default. The reason we don't get these requests often is because (a) people don't bother reading the default options and figuring out if they will work well; they just hand-configure them anyway (b) they don't know how other people do it and/or they don't read the TS docs, so they trust the default options and they just use camelCase all along since they created the project (c) they don't use enums at all (d) they don't find it meaningful to argue about configurable default options. But I think it's a nice improvement both for developer education and alignment with conventions.

**Comment by JoshuaKGoldberg** [comment]
In theory, per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

In practice, I think this is a nice and straightforward enough improvement for users that we can allow it regardless. cc @Josh-Cena

**Comment by JoshuaKGoldberg** [comment]
Oh, is this actually a breaking change? Do we need to wait for the next major? @typescript-eslint/triage-team 

I'm thinking it probably is, since some teams might be relying on _not_ allowing PascalCase?

**Comment by Josh-Cena** [comment]
Yeah it would be.

**Comment by kirkwaiblinger** [comment]
Oh, crap, I merged https://github.com/typescript-eslint/typescript-eslint/pull/11127 without realizing we wanted to treat it as breaking. PR to revert is up: https://github.com/typescript-eslint/typescript-eslint/pull/11447

**Comment by bradzacher** [comment]
I'm still _personally_ a -1 on this just cos:
- This has been the default for aaaages.
- It's a breaking change that has no fixer -- so it will cause disruption in users that do rely on this.
- The rule is frozen and we don't really want to make any changes.

**Comment by JoshuaKGoldberg** [comment]
> breaking change

My read is that we're just _adding_ to the list of allowed names - not making it more strict. As in, the rule would become more lenient by default?

> we don't really want to make any changes

In general yes, but I think making the rule less strict by default is a good step for making it less of a pain to deal with.

**Comment by bradzacher** [comment]
> My read is that we're just adding to the list of allowed names - not making it more strict.

No - adding the config as described in OP would enforce thst all enum members are named with PascalCase. Which is what I am describing as a severe breaking change that's too large to make.

We could achieve the behaviour you want of we instead use the following to allow either camel or pascal case.

```diff
+  {
+    selector: 'enumMember',
+    format: ['PascalCase', 'camelCase'],
+  },
```

But that's not what OP was suggesting.

If we want to allow both then sure - that's an entirely different story!

**Comment by JoshuaKGoldberg** [comment]
**OH** I was always looking for allowing _both_. As in https://github.com/typescript-eslint/typescript-eslint/issues/7879#issuecomment-1793436274. So I guess we're co-opting this issue from what the OP might have been proposing. 😄 

What is your stance on that entirely different story?

**Comment by bradzacher** [comment]
I don't love it cos it removed the nice consistency of the base config... But ultimately most people likely use the rule with a custom config so I doubt that too many people would run into the inconsistency that is introduced.

Loosening it and allowing inconsistency is sucky but okay enough given the pattern is prevalent.

**Comment by StyleShit** [comment]
I'm not in favor of loosening the rule, (my OCD is) agreeing with Brad here:

> it removed the nice consistency of the base config

And I'm -1 on this even more, assuming this is correct:

> most people likely use the rule with a custom config


I would've proposed to add this as a breaking change, but since this rule is frozen, I'm not sure this is something that we should touch

<sub>Side note: I personally don't like enums, but I _do_ prefer them to be PascalCase when they exist</sub>

**Comment by JoshuaKGoldberg** [comment]
Fair. At this point the issue has been open for 2.5-3 years with only 6 👍s. That's not a lot of positive signal. And the rule has been feature-frozen for a while. Closing out.

Thanks for filing, and thanks everyone who chipped in! As Brad said, the vast majority of users customize the rule - see https://typescript-eslint.io/rules/naming-convention/#options
